jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java

```diff
@@ -115,28 +115,67 @@ protected void doStop() throws Exception
     {
         if (LOG.isDebugEnabled())
             LOG.debug("Stopping {}", this);
-        CloseEndPoints close_endps = new CloseEndPoints();
-        submit(close_endps);
-        close_endps.await(getStopTimeout());
-        CloseSelector close_selector = new CloseSelector();
-        submit(close_selector);
-        close_selector.await(getStopTimeout());
-
-        super.doStop();
         
+        try
+        {
+            CloseAction close = new CloseAction();
+            submit(close);
+            if (getStopTimeout()>0)
+                close.await(getStopTimeout());
+        }
+        finally
+        {
+            try
+            {
+                super.doStop();
+            }
+            finally
+            {
+                try (Locker.Lock lock = _locker.lock())
+                {
+                    if (_selector!=null && _selector.isOpen())
+                        _selector.wakeup();
+
+                    _actions.stream().filter(CloseAction.class::isInstance).forEach(a->a.run());
+                    _actions.clear();
+                }
```

Comment:
Nope, we are calling `run()` while holding a lock.
We need to copy the actions we want to run, and then run them without the lock held.
Also, `a->a.run()` can be reduced to a method reference.

Reply:
I don't think so.  The actions are all intended to be run by the selector thread and not dispatched to arbitrary user code.  So it is fine for the lock to be held... in fact it is probably good.